### PR TITLE
Add the possibility of passing CMAKE flag to external projects

### LINKED ISSUE #342: Generate functionality to pass all variables to  External CMakes
### Preamble 

I am having serious problems compiling and running the SuperBuild. I did manage to have something working at some point by modifying and unknown amount of variables in the `cmake-gui` of `./build` and `./build/builds/XXXXXX/build`. At some point I managed to screw over it again and could not continue. @rijobro Suggested a better alternative to keep track of changes. To Have a bash script where I go passing variables as I manage to compile parts of it, something like:

```
mkdir -p build
cd build

cmake ../SIRF-SuperBuild \
-DBUILD_Gadgetron:BOOL=OFF \
 ...
```

This is great because I can go changing variables and turning on slowly the flags of libraries to build.

### Problem

This can not access flags that are not set by the SuperBuild Cmake. If someone wanted to change e.g. `BUILD_MATLAB_WRAPPER` inside `CCPi_Regularisation_Tookit/build`, the only way to do it is to go to that particular CMake, and change it. Unfortunately, this means to again, lose track of teh changes. Passing all variables to the main CMake sounds like a tedious and unmaintainable task

### Proposed Solution
(by @rijobro )
Make a macro `SetExtraArgs.cmake` that would parse input flags like: `-DCCPi_Regularisation-Toolkit_SET_VAR_BUILD_MATLAB_WRAPPER:BOOL=ON`, then parse it (regexp or otherwise, using `${proj}`) and extract the part after `_SET_VAR_`. Make a list and then pass it in the `CMAKE_ARGS` section of `External_CCPi-Regularisation-Toolkit.cmake` 

### Problems with this solution

Mostly that it exposes all variables (but encapsulation is not necessarily what we want here right?) and that we need to modify all the `External_*` files. 

I am happy to take upon this issue and happy to hear other options/solutions. 
Or is this something that its "just me" and we do not want it in the SuperBuild?

### PR DESCRIPTION
**Pull request by AnderBiguri**
Resolves #342. Allows to add particular project CMAKE args to the main call, together with any other flag. 

e.g.

```
cmake ./SIRF-SuperBuild -DUSE_SYSTEM_BOOST -DNIFTYPET_EXTRA_CMAKE_ARGS:STRING="-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON" -DSTIR_EXTRA_CMAKE_ARGS:STRING="-DBUILD_SWIG_MATLAB:BOOL=OFF -DDISABLE_ITK:BOOL=ON -DBUILD_MATLAB_WRAPPER:BOOL=ON"

```

This functionality has only been added to the following External projects:

- STIR
- Gadgetron
- SIRF
- NIFTYREG
- NIFTYPET
- CCPi-Regularisation-Toolkit
- TomoPhantom 

To add to a new project, simply:

1. Add `SetExternalProjectFlags(${proj})` to the relevant `External_XXX.cmake` file
2. Add `${${proj}_EXTRA_CMAKE_ARGS_LIST}` to the `CMAKE_ARGS` list. 

This PR requires https://github.com/CCPPETMR/SIRF-SuperBuild/pull/350

### CODE REVIEW

File: CMake/SetExternalProjectFlags.cmake [hunk 1]
```diff
@@ -0,0 +1,30 @@
+#========================================================================
+# Author: Ander Biguri
+# Copyright 2019 University College London
```
Comment (KrisThielemans) [inline CMake/SetExternalProjectFlags.cmake]:
```suggestion
# Copyright 2020 University College London
```

File: SuperBuild/External_Gadgetron.cmake [hunk 2]
```diff
@@ -35,6 +35,10 @@ set(externalProjName ${proj})
 
 SetCanonicalDirectoryNames(${proj})
 
+# Get any flag from the superbuild call that may be particular to this projects CMAKE_ARGS
+SetExternalProjectFlags(${proj})
+
+#message(FATAL_ERROR "${proj}_EXTRA_CMAKE_ARGS = \"${${proj}_EXTRA_CMAKE_ARGS}\"")
```
Comment (KrisThielemans) [inline SuperBuild/External_Gadgetron.cmake]:
```suggestion
```

File: CMake/SetExternalProjectFlags.cmake [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+#=========================================================================
+
+# This defines a macro that gets flags passed to the superbuild to modify particular project cmake flags and marks them as advanced.
+# Allows passing cmake flags to the superbuild to be reused by particular projects.
+
+macro(SetExternalProjectFlags proj)
+
+	set(${proj}_EXTRA_CMAKE_ARGS "" CACHE STRING "Optional extra CMake arguments to be appended to the flags set by the SuperBuild")
+	mark_as_advanced(${proj}_EXTRA_CMAKE_ARGS)
+	string(REPLACE " " ";" ${proj}_EXTRA_CMAKE_ARGS_LIST "${${proj}_EXTRA_CMAKE_ARGS}")
```
Comment (KrisThielemans) [inline CMake/SetExternalProjectFlags.cmake]:
```suggestion
	#string(REPLACE " " ";" ${proj}_EXTRA_CMAKE_ARGS_LIST "${${proj}_EXTRA_CMAKE_ARGS}")
```

File: CMake/SetExternalProjectFlags.cmake [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+# limitations under the License.
+#
+#=========================================================================
+
+# This defines a macro that gets flags passed to the superbuild to modify particular project cmake flags and marks them as advanced.
+# Allows passing cmake flags to the superbuild to be reused by particular projects.
+
+macro(SetExternalProjectFlags proj)
+
+	set(${proj}_EXTRA_CMAKE_ARGS "" CACHE STRING "Optional extra CMake arguments to be appended to the flags set by the SuperBuild")
```
Comment (KrisThielemans) [inline CMake/SetExternalProjectFlags.cmake]:
```suggestion
	set(${proj}_EXTRA_CMAKE_ARGS "" CACHE STRING "Optional extra CMake arguments to be appended to the flags set by the SuperBuild (use semi-colons for multiple arguments)")
```

**Comment by AnderBiguri** [comment]
@KrisThielemans  done

**Comment by AnderBiguri** [comment]
@KrisThielemans  I can not reopen this myself.

**Comment by KrisThielemans** [comment]
@AnderBiguri not sure if this affects any of your build scripts...

**Pull request merged**